crates/stages/stages/src/stages/execution.rs

```diff
@@ -149,9 +150,9 @@ impl<E> ExecutionStage<E> {
     /// Given that `start_block` changes with each checkpoint, it's necessary to inspect
     /// [`tables::AccountsTrie`] to ensure that [`super::MerkleStage`] hasn't
     /// been previously executed.
-    fn adjust_prune_modes<DB: Database>(
+    fn adjust_prune_modes(
         &self,
-        provider: &DatabaseProviderRW<DB>,
+        provider: &impl StatsReader,
```

Comment:
this is auto impl for & so we can accept impl here I believe
```suggestion
        provider: impl StatsReader,
```